### PR TITLE
Pivot RFC to generic framework for `box` and annotations

### PR DESCRIPTION
**Pull request by jakelishman**
This fleshes out a lot more details about how a generic `box` would work in Qiskit SDK, and how arbitrary downstream annotations might work in the context of an entire compiler framework, where annotation-creating passes may be injected inbetween standard Qiskit passes, or the user may write into their circuit, and the transpiler might want to verify before submission to the quantum computer.

The case study on how `pec-runtime` will use this isn't fleshed out - I mostly just moved it into a section at the end and left it.

### CODE REVIEW

File: 0022-circuit-block.md [hunk 1]
```diff
@@ -1,38 +1,285 @@
...
+This can be worked around by inserting explicit `id` gates into the box, or some other form of no-op operation, but these instructions can sometimes be treated by hardware as something other than a no-op, which might cause performance problems if the compiler fails to optimise them out or understand their intention.
+For example, IBM hardware has, in the past, treated the `id` instruction as a single-cycle delay on a single qubit, which is not precisely equal to a no-op for the purposes of scheduling, especially if the rest of the block has a natural duration of zero cycles (such as if it comprises only virtual phase update `rz` gates).
+We also do not currently have the concept of a no-op on classical data.
+To make the user intent clearer here, we propose adding a `QuantumCircuit.noop` to make the intent explicit:
+
+> [!NOTE]
+> *Bikeshedding potential*: do you prefer `QuantumCircuit.use` or `QuantumCircuit.noop` for this?
+> `use` is perhaps clearer for more users, but its name only makes sense in a control-flow builder context, whereas `noop` makes sense to appear in the root scope of a `QuantumCircuit`.
```
Comment (ihincks) [inline 0022-circuit-block.md:120]: No strong preference. I'm sure we and users would get used to noop.

File: 0022-circuit-block.md [hunk 2]
```diff
@@ -1,38 +1,285 @@
...
 
-The concept of a block of gates is fundamental to many quantum computing routines. For example, a block is a natural concept when reasoning about quantum errors because the error profile often depends on factors such as which gates are applied simultaneously. However, Qiskit does not provide a dedicated object to represent a block. Traditionally, users have worked around this by defining blocks *indirectly* using barriers. While this approach has been adequate for tasks like scheduling and transpilation, barriers can be hard to parse when pre-processing circuits in preparation for twirling or mitigation. Suboptimal handling of barriers in these post-processing steps can introduce significant slowdowns in some mitigation experiments (for example, it can unnecessarily increase the time required by the noise learning steps) and produce outcomes that may seem unintuitive to typical users.
+Introduce a way grouping of instructions to Qiskit SDK that can have downstream data attached, pass through transpilation, and up and down an execution stack.
```
Comment (SamFerracin) [inline 0022-circuit-block.md]:
```suggestion
Introduce a way of grouping instructions in Qiskit SDK that can have downstream data attached, pass through transpilation, and up and down an execution stack.
```

File: 0022-circuit-block.md [hunk 3]
```diff
@@ -1,38 +1,285 @@
...
+
+The RFC proposes to add a new `Box` instruction to Qiskit, whose semantics will be very similar to the `box` concept from OpenQASM 3.
+This `Box` should be able to pass through the Qiskit compiler (under certain limitations), be QPY serialisable, and be able to be transmitted up and down execution stacks.
+We will examine how this `Box` might also permit downstream "annotations", outside of the control of the Qiskit SDK compiler, and how the compiler will still be able to reason about such annotations.
+
+The nearest-term use cases for this are:
+
+* have a grouping construction to simplify stretch-based scheduling in Qiskit SDK
+* allow advanced users to specify the noise-learning and twirling blocks IBM Primitives error mitigation should use, and the per-block strategy
+* allow regular uses to see what blocks the IBM Primitives divided their circuit into for noise learning and twirling
```
Comment (SamFerracin) [inline 0022-circuit-block.md]:
```suggestion
* allow regular users to see what blocks the IBM Primitives divided their circuit into for noise learning and twirling
```

File: 0022-circuit-block.md [hunk 4]
```diff
@@ -1,38 +1,285 @@
...
+
+
+## `Box`: a one-off grouping of instructions
+
+As in OpenQASM 3, a `Box` will be:
+
+* a group of non-overlapping (but potentially nested) circuit instructions,
+* which introduces a new scope for declared variables,
+* which can have internal scheduling within itself, but it scheduled atomically within the circuit (i.e. all wires touched within the box are delayed to match the critical path length),
```
Comment (SamFerracin) [inline 0022-circuit-block.md]:
```suggestion
* which can have internal scheduling within itself, but is scheduled atomically within the circuit (i.e. all wires touched within the box are delayed to match the critical path length),
```

File: 0022-circuit-block.md [hunk 5]
```diff
@@ -1,38 +1,285 @@
...
+* and can have arbitrary backend- and SDK-agnostic annotations attached to it.
+
+A `box` can contain any other Qiskit circuit instruction.
+Aside from variable scoping concerns and the resolution of delay lengths, the execution of a circuit containing a `box` should be the same as the execution of the same circuit with the boxed instructions inlined into the circuit.
+
+### How to implement in Qiskit
+
+A `box` can be seen as the trivial case of a control-flow operation, equivalent to an `if (true)` block in a control-flow graph.
+Box, however, additionally has the semantics that is not a valid optimisation to simply remove the box.
+A `box` is a non-resuable grouping of instructions; it is not a function call that can be called multiple times (this is still a useful concept, it's just separate to `box` and not addressed here).
```
Comment (SamFerracin) [inline 0022-circuit-block.md]:
```suggestion
A `box` is a non-reusable grouping of instructions; it is not a function call that can be called multiple times (this is still a useful concept, it's just separate to `box` and not addressed here).
```

File: 0022-circuit-block.md [hunk 6]
```diff
@@ -1,38 +1,285 @@
...
+This might naturally extend to such a `BackendV3` providing the additional information that is needed for Qiskit SDK to safely transpile entire sampler and estimator pubs; the compilation of a pub _also_ implies further processing will be done by a quantum computer, just like an annotation remaining in a single circuit.
+
+
+## Case study: noise learning in `pec-runtime`
+
+We envision two main workflows that users will follow once `Box`es become available, one for regular users and another one for power users:
 
-We envision two main workflows that users will follow once `Block`s become available, one for regular users and another one for power users:
 1. The workflow for regular users:
     - Users initialize a `QuantumCircuit` without blocks, as they do today.
```
Comment (SamFerracin) [inline 0022-circuit-block.md]:
```suggestion
    - Users initialize a `QuantumCircuit` without boxes, as they do today.
```
